### PR TITLE
MSEARCH-69: Create system user on module initialization.

### PR DESCRIPTION
**Pull request by bohdan-suprun**

### CODE REVIEW

File: src/main/java/org/folio/search/service/systemuser/SystemUserService.java
Comment (hjiebsco) [inline src/main/java/org/folio/search/service/systemuser/SystemUserService.java:140]: This is a bit brute force. Maybe retrieving existing perms and compare with the config first.

**Review by hjiebsco — COMMENTED** [review]
- The `loginSystemUser` should happen when the token is needed the firs time. Currently it is only called by `prepareSystemUser` during `postTenant`. When adding new mod-search instance to a running env, I assume you will need to a place to call that method since the token is only cached in memory of individual instance.
- `permissions/mod-search.csv` file is empty

File: src/main/java/org/folio/search/model/SystemUser.java
Comment (bohdan-suprun) [reply src/main/java/org/folio/search/model/SystemUser.java:20]: Okapi url is needed to make a call to sub modules, there will be no execution context.

Comment (hjiebsco) [reply src/main/java/org/folio/search/model/SystemUser.java:20]: > Okapi url is needed to make a call to sub modules, there will be no execution context.

Okapi url is needed to make calls, but it should be a module level (app context) variable, not tied to a specific system user which is at tenant level.

File: README.md
Comment (bohdan-suprun) [reply README.md:61]: Added.

File: src/main/java/org/folio/search/service/systemuser/SystemUserService.java
Comment (bohdan-suprun) [reply src/main/java/org/folio/search/service/systemuser/SystemUserService.java]: Make sense, changed.

File: src/main/java/org/folio/search/service/systemuser/SystemUserService.java
Comment (bohdan-suprun) [reply src/main/java/org/folio/search/service/systemuser/SystemUserService.java]: Agree, changed.

File: src/main/java/org/folio/search/model/SystemUser.java
Comment (bohdan-suprun) [reply src/main/java/org/folio/search/model/SystemUser.java]: I have removed password from the class, all rest is used. 

> Okapi url is needed to make calls, but it should be a module level (app context) variable, not tied to a specific system user which is at tenant level.

Yes, but okapi token is not known at startup, so it is bound to folio-tenant scope in the base module.

Comment (hjiebsco) [reply src/main/java/org/folio/search/model/SystemUser.java]: Are we talking about `okapiUrl` or `okapiToken`? The first is at module level and the second is at tenant level.

Comment (bohdan-suprun) [reply src/main/java/org/folio/search/model/SystemUser.java]: Yes, sorry, I meant that *okapi URL* is not known at module start up, but only on tenant init or when a first request comes.

**Review by hjiebsco — COMMENTED** [review]
- Looks like `SystemUser` is only prepared during tenant enabling. Can you explain how it will be handled when adding a new mod-search instance to a running environment? I asked this question last time but I still do not see a mechanism to initialize `SystemUser` for  newly added mod-search instances.
- Also, maybe consider moving `AsyncFolioExecutionContext` to under the new `org.folio.search.service.context` package to be consistent.

**Comment by bohdan-suprun** [comment]
Hi @hjiebsco,

> Looks like SystemUser is only prepared during tenant enabling. Can you explain how it will be handled when adding a new mod-search instance to a running environment? 

Good catch! I found out that the current approach does not suitable for the multi-instance mode. There is no way but to persist the prepared user to DB. 

The lazy login can not be performed because we don't know OKAPI URL on that step.

Is it fine to persist following user data to DB:
* id;
* username (can be omitted if not fine);
* okapi URL
* okapi token (can be also omitted - in this case each instance will issue a separate token);
* tenant?

**Comment by hjiebsco** [comment]
> Hi @hjiebsco,
> 
> > Looks like SystemUser is only prepared during tenant enabling. Can you explain how it will be handled when adding a new mod-search instance to a running environment?
> 
> Good catch! I found out that the current approach does not suitable for the multi-instance mode. There is no way but to persist the prepared user to DB.
> 
> The lazy login can not be performed because we don't know OKAPI URL on that step.
> 
> Is it fine to persist following user data to DB:
> 
> * id;
> * username (can be omitted if not fine);
> * okapi URL
> * okapi token (can be also omitted - in this case each instance will issue a separate token);
> * tenant?

Or we can do this:
1. Make okapiUrl a config var or env var (just like system user)
2. During tenant enabling, create (if not exist) the system user and assign (or verify) perms
3. When module to module call happens, login as necessary to get okapiToken and cache it in memory

**Comment by bohdan-suprun** [comment]
Hi @hjiebsco, the system user is stored in DB now, please review.

**Comment by bohdan-suprun** [comment]
Hi @hjiebsco, @MikhailFokanov here is the ticket https://issues.folio.org/browse/MSEARCH-73

File: src/main/java/org/folio/search/service/systemuser/SystemUserService.java
Comment (hjiebsco) [inline src/main/java/org/folio/search/service/systemuser/SystemUserService.java:78]: If `tokenCache` has the token, I do not think you need `findSystemUser` first

**Review by hjiebsco — COMMENTED** [review]
I prefer cleaning up the implementation (removing the systemuser db at least) before merging but I will leave it up to you. I am done with reviewing this PR.

File: src/main/java/org/folio/search/service/systemuser/SystemUserService.java
Comment (bohdan-suprun) [reply src/main/java/org/folio/search/service/systemuser/SystemUserService.java]: Thanks, you're right.

**Review by hjiebsco — COMMENTED** [review]
LGTM except I prefer removing db part. I will defer @MikhailFokanov and @pfilippovEPAM to approve.

**Comment by bohdan-suprun** [comment]
> LGTM except I prefer removing db part. I will defer @MikhailFokanov and @pfilippovEPAM to approve.

@hjiebsco, this part has been moved to https://issues.folio.org/browse/MSEARCH-73, I have requested var adding in https://issues.folio.org/browse/FOLIO-3059

**Review by hjiebsco — APPROVED** [review]

**Review by pfilippovEPAM — APPROVED** [review]

**Pull request merged**
